sycl/doc/extensions/DataFlowPipes/data_flow_pipes.asciidoc

```diff
@@ -0,0 +1,650 @@
+
+
+[source,c++,FunctionCallExample,linenums]
+----
+void pipe_memcpy(int* dest, const int* src, size_t n) {
+  constexpr int N = 10;  // n <= N.  No checking here for simplicity
+  using mypipe = pipe<class local, int, N>;
+  for (size_t i = 0; i < n; ++i) mypipe::write(src[i]);
+  for (size_t i = 0; i < n; ++i) dest[i] = mypipe::read();
+}
+----
+
+
+For serialized calls to pipe_memcpy, within a kernel for example, all behaves as expected.
+
+[source,c++,SerializedFunctionCallExample,linenums]
+----
+pipe_memcpy(d1, s1, N);
+pipe_memcpy(d2, s2, N);
+----
+
+The same pipe type is reused by both of these calls, but the calls execute serially so the memcpy behavior is as expected.
+
+Once concurrency enters the picture, then `pipe_memcpy` potentially exhibits unexpected behavior because the pipe type declared locally to the function is identical/common across the calls.  This can be exposed through calls from different kernels that happen to be executing concurrently, or by calls from different work-items in the same kernel.  The pipe therefore becomes a resource that is identical/common across calls to `pipe_memcpy`, and requires either uniquification between invocations, or handling similar to concurrent accesses to a shared memory resource.  This is the expected and desired behavior, but if users intend for the pipe to be privatized per invocation, then they must explicitly code for that.
+
+Note that calls from the same kernel and same work-item which happen to be inlined and otherwise optimized to execute "concurrently" will not exhibit incorrect behavior - sequential consistency within a work-item is still required.  The pipe accesses have side effects, and each pipe is conceptually in it's own address space that doesn't alias with any others.  Within a single work-item, the compiler is not free to arbitrarily reorder operations with side effects.
+
+Uses of pipes declared within a function call that is invoked in a concurrent way are not common, but are legal.
+
+Users can privatize the pipe type by templating the function, if they do require independent pipes for concurrent calls to the function.  For example:
+
+[source,c++,TemplatedFunctionCallExample,linenums]
+----
+template <int id>
+void pipe_memcpy(int* dest, const int* src, size_t n) {
+  constexpr int N = 10;  // n <= N.  No checking here for simplicity
+  using mypipe = pipe<class local, int, N>;
+  for (size_t i = 0; i < n; ++i) mypipe::write(src[i]);
+  for (size_t i = 0; i < n; ++i) dest[i] = mypipe::read();
+}
+
+...
+
+pipe_memcpy<1>(d1, s1, N);
+pipe_memcpy<2>(d2, s2, N);
+----
+
+The above example function `pipe_memcpy()` could alternatively be templated on the pipe type (first template argument of the pipe class specialization), amongst other possibilities.
+
+Automated mechanisms are possible to provide uniquification across calls, and could be exposed through a wrapper or library.
```

Comment:
Interestingly, all this section makes a different assumptions about pipes from the one in Xilinx Vivado HLS, where some people insist about having arrays of pipes. I am unsure array of pipes is a great idea anyway, since a similar data structure can be provided as a meta-programmed templated library to hold different pipe data types and dispatch read/write operations from a runtime index...